sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/UnsafeHybridAggregationIterator.scala

```diff
@@ -0,0 +1,398 @@
+
+  require(groupingKeyAttributes.nonEmpty)
+
+  ///////////////////////////////////////////////////////////////////////////
+  // Unsafe Aggregation buffers
+  ///////////////////////////////////////////////////////////////////////////
+
+  // This is the Unsafe Aggregation Map used to store all buffers.
+  private[this] val buffers = new UnsafeFixedWidthAggregationMap(
+    newBuffer,
+    StructType.fromAttributes(allAggregateFunctions.flatMap(_.bufferAttributes)),
+    StructType.fromAttributes(groupingKeyAttributes),
+    TaskContext.get.taskMemoryManager(),
+    SparkEnv.get.shuffleMemoryManager,
+    1024 * 16, // initial capacity
+    SparkEnv.get.conf.getSizeAsBytes("spark.buffer.pageSize", "64m"),
+    false // disable tracking of performance metrics
+  )
+
+  override protected def newBuffer: UnsafeRow = {
+    val bufferSchema = allAggregateFunctions.flatMap(_.bufferAttributes)
+    val bufferRowSize: Int = bufferSchema.length
+
+    val genericMutableBuffer = new GenericMutableRow(bufferRowSize)
+    val unsafeProjection =
+      UnsafeProjection.create(bufferSchema.map(_.dataType))
+    val buffer = unsafeProjection.apply(genericMutableBuffer)
+    initializeBuffer(buffer)
+    buffer
+  }
+
+  ///////////////////////////////////////////////////////////////////////////
+  // Methods and variables related to switching to sort-based aggregation
+  ///////////////////////////////////////////////////////////////////////////
+  private[this] var sortBased = false
+
+  private[this] var sortBasedAggregationIterator: SortBasedAggregationIterator = _
+
+  // The value part of the input KV iterator is used to store original input values of
+  // aggregate functions, we need to convert them to aggregation buffers.
+  private def processOriginalInput(
+      firstKey: UnsafeRow,
+      firstValue: InternalRow): KVIterator[UnsafeRow, UnsafeRow] = {
+    new KVIterator[UnsafeRow, UnsafeRow] {
+      private[this] var isFirstRow = true
+
+      private[this] var groupingKey: UnsafeRow = _
+
+      private[this] val buffer: UnsafeRow = newBuffer
+
+      override def next(): Boolean = {
+        initializeBuffer(buffer)
+        if (isFirstRow) {
+          isFirstRow = false
+          groupingKey = firstKey
+          processRow(buffer, firstValue)
+
+          true
+        } else if (inputKVIterator.next()) {
+          groupingKey = inputKVIterator.getKey()
+          val value = inputKVIterator.getValue()
+          processRow(buffer, value)
+
+          true
+        } else {
+          false
+        }
+      }
+
+      override def getKey(): UnsafeRow = {
+        groupingKey
+      }
+
+      override def getValue(): UnsafeRow = {
+        buffer
+      }
+
+      override def close(): Unit = {
+        // Do nothing.
+      }
+    }
+  }
+
+  // The value of the input KV Iterator has the format of groupingExprs + aggregation buffer.
+  // We need to project the aggregation buffer out.
+  private def projectInputBufferToUnsafe(
```

Comment:
we should just remove this function and inline it. We don't want an extra iterator overhead to process the rows.

Each iterator actually adds a lot of overhead, and here it doesn't buy you any code reduction (on the contrary it increases complexity due to the extra abstraction).